### PR TITLE
Propagation angles

### PR DESCRIPTION
**Pull request by isildur7**
Added angled propagation (only to exact_propagate at the moment). Also added support for non-zero padding values.
for propagation, we shift the kernel by (kx, ky). Will edit the comment to add more details about this.

### CODE REVIEW

File: src/chromatix/functional/propagation.py [hunk 1]
```diff
@@ -104,6 +114,8 @@ def exact_propagate(
     n: float,
     *,
     N_pad: int,
+    pad_value: float = 0,
```
Comment (diptodip) [inline src/chromatix/functional/propagation.py:117]: Let's call it cval to make it consistent with numpy/scipy conventions

File: src/chromatix/functional/propagation.py [hunk 2]
```diff
@@ -115,21 +127,27 @@ def exact_propagate(
...
     # Calculating propagator
+    if kykx is None:
+        kykx = jnp.zeros(2)
     f = []
     for d in range(field.dx.size):
         f.append(jnp.fft.fftfreq(field.shape[1] + N_pad, d=field.dx[..., d].squeeze()))
     f = jnp.stack(f, axis=-1)
     fx, fy = rearrange(f, "h c -> 1 h 1 c"), rearrange(f, "w c -> 1 1 w c")
-    kernel = 1 - (field.spectrum / n) ** 2 * (fx**2 + fy**2)
+    kernel = 1 - (field.spectrum / n) ** 2 * ((fx - kykx[1]) ** 2 + (fy - kykx[0]) ** 2)
```
Comment (diptodip) [inline src/chromatix/functional/propagation.py:145]: Looks good

File: src/chromatix/functional/propagation.py [hunk 3]
```diff
@@ -144,70 +162,71 @@ def exact_propagate(
...
-    loop_axis: Optional[int] = None,
-) -> Field:
+def calculate_padding_transform(
+    height: int, spectrum: float, dx: float, z: float
+) -> int:
+    """
+    Automatically calculate the padding required for transform propagation.
+
+    Args:
+        height: Shape of the field
```
Comment (diptodip) [inline src/chromatix/functional/propagation.py:172]: This should say height of the field

And we should have a TODO that this only works on square fields for now

File: src/chromatix/functional/propagation.py [hunk 4]
```diff
@@ -144,70 +162,71 @@ def exact_propagate(
...
+
+def calculate_padding_transfer(
+    height: int, spectrum: float, dx: float, z: float
+) -> int:
+    """
+    Automatically calculate the padding required for transfer propagation.
 
     Args:
-        field: ``Field`` to be propagated.
+        height: Shape of the field
```
Comment (diptodip) [inline src/chromatix/functional/propagation.py:195]: Same here

File: src/chromatix/utils/utils.py [hunk 5]
```diff
@@ -104,13 +104,15 @@ def next_order(val: int) -> int:
     return int(2 ** np.ceil(np.log2(val)))
 
 
-def center_pad(u: jnp.ndarray, pad_width: Sequence[int]) -> jnp.ndarray:
+def center_pad(
+    u: jnp.ndarray, pad_width: Sequence[int], pad_value: float = 0
```
Comment (diptodip) [inline src/chromatix/utils/utils.py:108]: Use cval here

File: src/chromatix/utils/utils.py [hunk 6]
```diff
@@ -104,13 +104,15 @@ def next_order(val: int) -> int:
...
+) -> jnp.ndarray:
     """
     Symmetrically pads ``u`` with lengths specified per axis in ``n_padding``,
     which should be iterable and have the same size as ``u.ndims``.
     """
     pad = [(n, n) for n in pad_width]
-    return jnp.pad(u, pad)
+    return jnp.pad(u, pad, constant_values=pad_value)
```
Comment (diptodip) [inline src/chromatix/utils/utils.py:115]: And here

File: tests/test_propagate.py [hunk 7]
```diff
@@ -88,25 +88,3 @@ def test_exact_propagation():
     # Exact is a bit worse here since it requires a lot of padding.
     # TODO: Find better test case.
     assert rel_error < 1e-2
-
-
-def test_propagate():
```
Comment (diptodip) [inline tests/test_propagate.py:93]: Why is this gone?

**Review by diptodip — CHANGES_REQUESTED** [review]

File: tests/test_propagate.py (hunk 7 above)
Comment (isildur7) [reply tests/test_propagate.py:93]: Because we got rid of the `propagate` method since it was not jitable. People need to call one of the three propagate methods directly with the padding calculator completely removed from the loop.

File: src/chromatix/functional/propagation.py [hunk 8]
```diff
@@ -115,21 +127,28 @@ def exact_propagate(
...
+            ConcretizationError will arise when traced!). Use padding calculator
+            utilities from ``chromatix.functional.propagation`` to calculate the
+            padding.
+        cval: The value for the padding, 0 by default
+        kykx: If provided, defines the orientation of the propagation. Should be an
+            array of shape `[2,]` in the format [ky, kx].
     """
     # Calculating propagator
+    if kykx is None:
+        kykx = jnp.zeros(2)
```
Comment (GJBoth) [inline src/chromatix/functional/propagation.py:140]: Default should be zeros instead of. It should also be `jnp.zeros((2, ))`.

File: src/chromatix/utils/utils.py (hunk 5 above)
Comment (isildur7) [reply src/chromatix/utils/utils.py:108]: done

File: src/chromatix/utils/utils.py (hunk 6 above)
Comment (isildur7) [reply src/chromatix/utils/utils.py:115]: done

File: src/chromatix/functional/propagation.py (hunk 8 above)
Comment (isildur7) [reply src/chromatix/functional/propagation.py:140]: done

File: src/chromatix/functional/propagation.py (hunk 3 above)
Comment (isildur7) [reply src/chromatix/functional/propagation.py:173]: done

File: src/chromatix/functional/propagation.py (hunk 4 above)
Comment (isildur7) [reply src/chromatix/functional/propagation.py:196]: done

**Review by diptodip — APPROVED** [review]
